Sources/NIO/BaseSocketChannel.swift

```diff
@@ -443,17 +443,18 @@ class BaseSocketChannel<T: BaseSocket>: SelectableChannel, ChannelCore {
         } catch let err {
             // If there is a write error we should try drain the inbound before closing the socket as there may be some data pending.
             // We ignore any error that is thrown as we will use the original err to close the channel and notify the user.
-            if readIfNeeded0() {
+            if self.readIfNeeded0() {
                 assert(self.lifecycleManager.isActive)
 
                 // We need to continue reading until there is nothing more to be read from the socket as we will not have another chance to drain it.
                 while let read = try? readFromSocket(), read == .some {
-                    assert(self.lifecycleManager.isActive)
-                    pipeline.fireChannelReadComplete()
+                    if self.lifecycleManager.isActive {
+                        pipeline.fireChannelReadComplete()
```

Comment:
No race here?

Reply:
No, this is thread-safe.

I think a more important question is why we fire `readComplete` more than once. I feel like we shouldn't be doing that: we should probably move the `fireChannelReadComplete` outside the `while` loop.

Reply:
@Lukasa I have no idea either why we're doing that but left it because it was that way. But let me just change it

Reply:
@vlm thread-safety is fine here because we're on the Channel's event loop. So (unless we call out), the value of `self.lifecycleManager.isActive` can't change. So if it's still active after `readFromSocket` returns, it'll remain active until we return or call out again.

Reply:
@Lukasa done